File: 25.k-个一组翻转链表.py
#
# @lc app=leetcode.cn id=25 lang=python3
#
# [25] K 个一组翻转链表
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        dummy = ListNode(None)
        dummy.next = head
        pre = end = dummy
        while end.next:
            for _ in range(k):
                end = end.next
                if not end:
                    # the remaining number of nodes less than k
                    return dummy.next
            start, tmp = pre.next, end.next
            end.next = None     # cut
            pre.next = self.reverse(start)
            start.next = tmp
            pre = start
            end = pre
        return dummy.next

    @staticmethod
    def reverse(head: ListNode):
        pre, cur = None, head
        while cur:
            tmp = cur.next
            cur.next = pre
            pre = cur
            cur = tmp
        return pre

# @lc code=end

File: 25.k-个一组翻转链表.java
/*
 * @lc app=leetcode.cn id=25 lang=java
 *
 * [25] K 个一组翻转链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        ListNode pre = dummy, end = dummy;
        while (end != null) {
            // move pointer to node of interest every k steps
            for (int i = 0; i < k; ++i) {
                end = end.next;
                if (end == null)  return dummy.next;    // keep order if step less than k
            }
            
            // reverse and reconnect
            ListNode start = pre.next, tmp = end.next;
            end.next = null;
            pre.next = reverse(start);
            start.next = tmp;

            // move pointers
            pre = start;
            end = pre;
        }
        return dummy.next;
    }

    private static ListNode reverse(ListNode head) {
        ListNode pre = null, cur = head;
        while (cur != null) {
            ListNode tmp = cur.next;
            cur.next = pre;
            pre = cur;
            cur = tmp;
        }
        return pre;
    }
}
// @lc code=end

